# HG changeset patch
# User Benjamin Lindner <lindnerb@users.sourceforge.net>
# Date 1248531605 -7200
# Node ID b308b2e12f04b12872adf020939e763e48b50771
# Parent  6ddb81224550376f80118d7148f424ae8ebc9f4e
determine correct image bitwidth in __magick_read__.cc

diff -r 6ddb81224550 -r b308b2e12f04 src/ChangeLog
--- a/src/ChangeLog	Sat Jul 25 16:17:27 2009 +0200
+++ b/src/ChangeLog	Sat Jul 25 16:20:05 2009 +0200
@@ -1,3 +1,8 @@
+2009-07-17  Benjamin Lindner <lindnerb@users.sourceforge.net>
+
+	* DLD-FUNCTIONS/__magick_read__.cc (F__magick_read__):
+	Determine correct number of bits required when reading images.
+
 2009-07-16  John W. Eaton  <jwe@octave.org>
 
 	* graphics.cc (get_array_limits): Require min_pos value to be
diff -r 6ddb81224550 -r b308b2e12f04 src/DLD-FUNCTIONS/__magick_read__.cc
--- a/src/DLD-FUNCTIONS/__magick_read__.cc	Sat Jul 25 16:17:27 2009 +0200
+++ b/src/DLD-FUNCTIONS/__magick_read__.cc	Sat Jul 25 16:20:05 2009 +0200
@@ -416,11 +416,15 @@
   else
     {
       unsigned int depth = imvec[0].modulusDepth ();
-      int i = 0;
-      while (depth >>= 1)
-        i++;
-      depth = 1 << i;
-
+      if (depth > 1)
+	{
+	  --depth;
+	  int i = 1;
+	  while (depth >>= 1)
+            i++;
+	  depth = 1 << i;
+	}
+      
       switch (depth)
         {
         case 1:
